### PR TITLE
Add Paragraph Delimited Mappings Extended

### PR DESCRIPTION
**Pull request by Asbestosstar**
I have added the format Paragraph Delimited Mappings Extended https://pagure.io/FeatureCreep/MappingExFormat with the exception of the Include(inheritance structure) and AccessFlag parts which do not have equivalents in other mapping formats.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -10,7 +10,7 @@ Progress: View [the project](https://github.com/orgs/MaxPixelStudios/projects/1)
 For usage and more information, visit [the wiki](https://github.com/MaxPixelStudios/MinecraftDecompiler/wiki)(Currently in progress)
 
 ## Features
-1. Deobfuscate Minecraft using Proguard(Official) mappings, SRG, TSRG, CSRG and Tiny mappings
+1. Deobfuscate Minecraft using Proguard(Official) mappings, SRG, TSRG, CSRG, ParagraphDelimitedMappingsExtended and Tiny mappings
```
Comment (XiaoPangxie732) [inline README.md:13]: new mappings should be placed last, like:
Deobfuscate Minecraft using Proguard(Official) mappings, SRG, TSRG, CSRG, Tiny, and ParagraphDelimitedMappingsExtended mappings

File: modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/format/MappingFormats.java [hunk 2]
```diff
@@ -178,4 +202,4 @@ public static boolean registerMappingFormat(MappingFormat<?, ?> format) {
     public static Set<String> getFormatNames() {
         return ObjectSets.unmodifiable(MAPPING_FORMATS.keySet());
     }
-}
\ No newline at end of file
+}
```
Comment (XiaoPangxie732) [inline modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/format/MappingFormats.java:205]: what happened here?

File: modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/processor/MappingProcessors.java [hunk 3]
```diff
@@ -478,6 +480,183 @@ private static int processTree2(int index, int size, ObjectList<String> content,
 
         private static void error() {
             throw new IllegalArgumentException("Is this a Tiny v2 mapping file?");
-        }
+        }
```
Comment (XiaoPangxie732) [inline modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/processor/MappingProcessors.java:483]: please remove suspicious white spaces

File: modules/mapping-format-detector/src/main/java/cn/maxpixel/mcdecompiler/mapping/detector/DefaultDetectionUnit.java [hunk 4]
```diff
@@ -26,8 +27,9 @@ public boolean canDetectHeader() {
         String s = contents.get(0);
         if (s.startsWith("PK: ") || s.startsWith("CL: ") || s.startsWith("FD: ") || s.startsWith("MD: ")) return Pair.of(MappingFormats.SRG, Percentage.NINETY_NINE);
         else if (s.endsWith(":")) return Pair.of(MappingFormats.PROGUARD, Percentage.NINETY);
+        else if (s.contains("¶") || s.contains("\\u00B6")) return Pair.of(MappingFormats.PDME, Percentage.NINETY);
```
Comment (XiaoPangxie732) [inline modules/mapping-format-detector/src/main/java/cn/maxpixel/mcdecompiler/mapping/detector/DefaultDetectionUnit.java:30]: you should only choose one, either header detection or content detection

File: modules/mapping-format-detector/src/main/java/cn/maxpixel/mcdecompiler/mapping/detector/DefaultDetectionUnit.java [hunk 5]
```diff
@@ -26,8 +27,9 @@ public boolean canDetectHeader() {
...
         s = contents.get(1);
         if (s.startsWith("\t")) return Pair.of(MappingFormats.TSRG_V1, Percentage.FIFTY);
         else return Pair.of(MappingFormats.CSRG, Percentage.FIFTY);
     }
-}
\ No newline at end of file
+}
```
Comment (XiaoPangxie732) [inline modules/mapping-format-detector/src/main/java/cn/maxpixel/mcdecompiler/mapping/detector/DefaultDetectionUnit.java:35]: also what happened here

File: modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/generator/MappingGenerators.java
Comment (XiaoPangxie732) [inline modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/generator/MappingGenerators.java:1]: The code style should match this project

File: modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/processor/MappingProcessors.java [hunk 6]
```diff
@@ -478,6 +480,183 @@ private static int processTree2(int index, int size, ObjectList<String> content,
...
+		
+		private String putDocs(String[] row, PairedMapping mapping, boolean req) {
+			if (row.length > 5) {
+				if (req) {mapping.getComponent(Documented.class).setContentString(row[5]);}
+				return row[5];
+			}
+			return null;
+		}
+	};
+
```
Comment (XiaoPangxie732) [inline modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/processor/MappingProcessors.java:659]: I will look into this on the weekend

**Review by XiaoPangxie732 — CHANGES_REQUESTED** [review]

**Comment by XiaoPangxie732** [comment]
The processor and the generator part are big, so I will review them on the weekend

**Comment by XiaoPangxie732** [comment]
by the way, I'm curious about how you type `¶`

File: modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/format/MappingFormats.java (hunk 2 above)
Comment (Asbestosstar) [reply modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/format/MappingFormats.java:205]: I think that is just from the added lines for the new format

File: modules/mapping-format-detector/src/main/java/cn/maxpixel/mcdecompiler/mapping/detector/DefaultDetectionUnit.java (hunk 5 above)
Comment (Asbestosstar) [reply modules/mapping-format-detector/src/main/java/cn/maxpixel/mcdecompiler/mapping/detector/DefaultDetectionUnit.java:35]: just from adding lines

File: modules/mapping-format-detector/src/main/java/cn/maxpixel/mcdecompiler/mapping/detector/DefaultDetectionUnit.java (hunk 4 above)
Comment (Asbestosstar) [reply modules/mapping-format-detector/src/main/java/cn/maxpixel/mcdecompiler/mapping/detector/DefaultDetectionUnit.java]: ok i chose content

File: modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/processor/MappingProcessors.java (hunk 3 above)
Comment (Asbestosstar) [reply modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/processor/MappingProcessors.java]: should be gone

File: modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/generator/MappingGenerators.java
Comment (XiaoPangxie732) [reply modules/mapping-api/src/main/java/cn/maxpixel/mcdecompiler/mapping/generator/MappingGenerators.java:1]: well, I'll take care of it

**Review by XiaoPangxie732 — APPROVED** [review]

**Comment by XiaoPangxie732** [comment]
wait what

**Comment by XiaoPangxie732** [comment]
1

**Review by XiaoPangxie732 — APPROVED** [review]
I'll merge these for now. More stuff will be done after this PR is merged.

**Pull request merged**